macros/sql_gen/gen_joined_metrics_cte.sql

```diff
@@ -139,11 +139,21 @@
     {%- else %}
         join_metrics__{{previous_cte_number}}.*
     {%- endif %}
-    {%- for metric in metric_tree.ordered_expression_set %}
+    {%- for metric in metric_tree.expression_set %}
         {%- if metric_tree.ordered_expression_set[metric] == cte_number %}
-        ,({{metrics_dictionary[metric].expression | replace(".metric_value","")}}) as {{metrics_dictionary[metric].name}}
+            {%- set metric_definition = metrics_dictionary[metric].expression %}
+            {%- if "/" in metric_definition -%}
+                {%- set split_division_metric = metric_definition.split('/') -%}
+                {%- set dividend = split_division_metric[0] -%}
+                {%- set divisors = split_division_metric[1:] | list -%}
+                {%- set expression = dividend ~ " / nullif(" ~ divisors | join(", 0) / nullif(") ~ ", 0)" -%}
```

Comment:
Oh dear. This scares me. Is there anything we could be doing to require these to be explicitly separated out into numerator and denominator at definition-time, instead of us trying to detect division at execution-time?

I'm also not entirely clear on what it's doing - can you add a comment showing the before and after?

Reply:
happy to add a comment here! where would you expect us to do this parsing if not here? in the creation of the `metric_dictionary`?

Reply:
as for what it's doing, it's parsing the full expression string into chunks, separated by the division sign. The first element in the list is treated as the numerator (dividend) and everything thereafter is a divisor. The outputted expression is then 

`dividend / nullif(divisor_1, 0) / nullif(divisor_2 ,0) ... nullif(divisor_n ,0)` 

just in case folks want to do a whole lot of division

Reply:
@dave-connors-3 without digging into your logic, what would happen if I had `metric a / (metric b / metric c)`?

Reply:
nothing good I'll tell ya that much

`metric a / nullif((metric b, 0) / nullif(metric c),0)` aka An Error

Reply:
Dang. Could we parse each division statement into a [DIV0](https://docs.snowflake.com/en/sql-reference/functions/div0.html) instead?

So this would be `DIV0(metric a, DIV0(metric b, metric c))`

Reply:
Looks like bigquery has [safe_divide](https://cloud.google.com/bigquery/docs/reference/standard-sql/mathematical_functions#safe_divide).

Looks like redshift and postgres don't have this native functionality.

Reply:
> what would happen if I had `metric a / (metric b / metric c)`?

Yeah this was what scared me, but I couldn't think of a real-world example and didn't want to scaremonger. 



> Looks like bigquery has [safe_divide](https://cloud.google.com/bigquery/docs/reference/standard-sql/mathematical_functions#safe_divide).

https://github.com/dbt-labs/dbt-utils/issues/679 will bring a cross-db implementation of this, but I doubt it'll handle being given a metric object...

> where would you expect us to do this parsing if not here

I would _prefer_ that we not be parsing at all, and it would be structured yaml: 
```yaml
# models/marts/product/schema.yml
version: 2

models:
 - name: dim_customers
   ...

metrics:
  - name: average_revenue_per_customer
    label: Average Revenue Per Customer
    description: "The average revenue received per customer"

    type: derived
    expression:
       numerator: {{metric('total_revenue')}}
       denominator: {{metric('count_of_customers')}}
    ...
```
I.e. `expression` would be able to be either a string (such as `{{ metric('new_customers') }} + {{ metric('existing_customers') }}`) or a dict with numerator/denominator

@jtcohen6: "We are locking down rc1 of dbt 1.3 next week, no more big changes! 
@joellabes: Hold my beer

Reply:
also I'll just leave these here: 
https://docs.exasol.com/db/latest/sql_references/functions/alphabeticallistfunctions/div.htm
https://www.w3schools.com/sql/func_mysql_div.asp

Reply:
@joellabes I don't think that fixes our issue though? Someone could still put `metric a / (metric b / metric c) in the denominator given that we support arbitrary sql statements. At this point its making me think - can we put some validation to make sure there is only one divide in the entire statement and say we don't support multiple? If a user **really** needed to do the above pattern then they could do
- metric b / metric c == metric d
- metric a/ metric d

Reply:
Yeah I don't think division on division is allowed.

Reply:
Isn't allowed as in we should prevent that behavior?

Reply:
Sorry that was unclear. I meant we shouldn't include it as a supported paradigm - we won't block you if we see two `/`s in an expression (or an additional `/` in a `denominator` field 🥺), but if it breaks then you're on your own. 

We document it as not recommended, but if you make it work (probably wrapping everything in `nullif`s yourself) then more power to you for sneaking one past the goalie.

Reply:
For the sake of consistent behavior, we're going to punt on this and @dave-connors-3 will write up an issue with all he learned in the development process. In the future we'll want a more holistic way to address this but for now this helps ensure consistent behavior.